src/scheduler/server_info.cpp

```diff
@@ -2265,72 +2205,70 @@ dup_server_info(server_info *osinfo)
 	/* the jobs are not dupped when we dup the nodes, so we need to copy
 	 * the node's job arrays now
 	 */
-	for (i = 0; osinfo->nodes[i] != NULL; i++)
-		nsinfo->nodes[i]->job_arr =
-			copy_resresv_array(osinfo->nodes[i]->job_arr, nsinfo->jobs);
-
-	nsinfo->num_parts = osinfo->num_parts;
-	if (osinfo->nodepart != NULL) {
-		nsinfo->nodepart = dup_node_partition_array(osinfo->nodepart, nsinfo);
-		if (nsinfo->nodepart == NULL) {
-			free_server(nsinfo);
-			return NULL;
+	for (int i = 0; osinfo.nodes[i] != NULL; i++)
+		nodes[i]->job_arr =
+			copy_resresv_array(osinfo.nodes[i]->job_arr, jobs);
+
+	num_parts = osinfo.num_parts;
+	if (osinfo.nodepart != NULL) {
+		nodepart = dup_node_partition_array(osinfo.nodepart, this);
+		if (nodepart == NULL) {
+			free_server_info();
```

Comment:
I don't think this will work.  Even though we threw an exception, we'll still leak.  When the caller does a foo = new server_info(*osinfo), they'll still get a new, but empty sinfo back.  They'll have to call delete on it.  If they're calling delete on it, then you don't have to free everything here.